### PR TITLE
[pigeon] Creates new Generator classes for each language

### PR DESCRIPTION
**Pull request by tarrinneal**
Hopefully less controversial version of https://github.com/flutter/packages/pull/2985.

Creates new `Generator` superclass and new language specific subclasses.

Also renames former `Generator` classes to `GeneratorAdapter` and updates tests accordingly.

Initial set up for https://github.com/flutter/flutter/issues/117416

### CODE REVIEW

File: packages/pigeon/lib/generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/generator.dart:28]: I think that the languageOptions root and sink should be parameters to the generate method.  Making them instance variables makes Generator stateful which severs the link between the input data and its consumption.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (gaaclarke) [inline packages/pigeon/lib/pigeon_lib.dart:360]: If you make Adapter a concrete class and Generator stateless like I mentioned in the other PR you'll find that there is a lot of boilerplate and duplication you could eliminate.

Eventually the list of adapters could look like this:
```dart
adapters = <Adapter>[
  Adapter(const CppHeaderGenerator()),
  Adapter(const CppSourceGenerator()),
  Adapter(const DartGenerator()),
];
```

**Comment by tarrinneal** [comment]
@gaaclarke I'd love to get another review on the latest commits. Specifically the way the new options are handled.

**Comment by gaaclarke** [comment]
> @gaaclarke I'd love to get another review on the latest commits. Specifically the way the new options are handled.

I think that looks good.  Originally I was concerned that people might mistakenly open file handles to those paths and fiddle around with them instead of using the provided sink.  What we found was those file paths were sometimes necessary for generating code.  We could probably clean it up with having specific options for the different generators (ie CppHeaderOptions, CppSourceOptions).  That's probably not worth the effort though.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:25]: Nit: this was more readable before, with the trailing comma.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:37]: ?

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:69]: Don't we only want one generator for C++ and ObjC? We should set this up the way we want it now, because this is new public API and changing it later would be breaking (which isn't a huge deal for Pigeon, but there's no reason to set ourselves up to have to do it.

Dart outputs multiple files from the same generator, so presumably we can do the same here.

File: packages/pigeon/lib/generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator.dart:7]: "A superclass"

File: packages/pigeon/lib/generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator.dart:9]: We probably won't remember to change this later; better to describe the intended structure rather than future work. E.g., "This provides the structure that is common across generators for different languages."

File: packages/pigeon/lib/generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator.dart:12]: An abstract class doesn't need a constructor.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:360]: `GeneratorAdapter` would be a better name; "Adapter" could be adapting anything.

Also, can this be private?

Also, this is a breaking change, so needs to be versioned as such. Everything in Pigeon is directly in `lib`, so everything that doesn't start with `_` is considered public.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/pigeon/lib/pigeon_lib.dart
Comment (tarrinneal) [reply packages/pigeon/lib/pigeon_lib.dart:360]: This is currently public for testing purposes. I'd be open to making adjustments to change this though. @gaaclarke mentioned adding the @public_for_testing tag on these classes.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:28]: Just `headerOut` since it's in `CppOptions`.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:42]: Having this and `header` is confusing. How about `headerOut` (the `cpp` is redundant with the options structure name) for this, and `headerIncludePath` for what's currently called `header`?

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:86]: Why are you passing both the options and a sub-element of the options?

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:30]: `Path` is inconsistent with the same field in the C++ options; all of them should either use the `Path` suffix or not, consistently. (I would vote with, for added clarity).

`dart` is redundant with the struture, and doesn't actually distinguish from the test code which is also Dart; I would call it `sourceOutPath`.

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:86]: Do we actually want a separate generator for this?

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:79]: Having this be a field still means we need to instantiate two generators. Couldn't this be a parameter to `generate` instead, so that a single generator can `generate` twice?

Comment (tarrinneal) [reply packages/pigeon/lib/cpp_generator.dart:79]: It could be, but that would be added to all other generators as well. I played around with a few options that all were worse than this. If there's a piece of information I'm missing to make this work, I'd be happy to change it.

If it's worth adding an unused parameter onto all of the other generators, that's fine too. If that is what's best.

Comment (stuartmorgan) [reply packages/pigeon/lib/cpp_generator.dart]: Another option would be to change `generate` to take a map of filetype->sink. Then there's a single generation call, which gets all of the relevant sinks.

Comment (tarrinneal) [reply packages/pigeon/lib/cpp_generator.dart]: That's not a bad idea. Should I eliminate the extra Generator adapters also? That's a somewhat larger process. If that's the endgoal though, I might as well make that happen now.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:73]: Nit: Remove "header".

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:37]: This comment is a copy/paste of the test path comment.

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:86]: Optional: Why not make this a file type as well so the structure is the same?

File: packages/pigeon/lib/kotlin_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/kotlin_generator.dart:77]: You might want to add an assert that the type is `source` in all the languages where that's the only expected value.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:364]: This is just a restatement of the code, so isn't a useful comment on its own. Let's combine it with the substance below to make the comment just:

```
/// A list of file types the generator should create.
```

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:369]: Nit for all of these comments: we should follow [Dart style](https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph) about the first sentence.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:384]: Can we make these named rather than positional? Having positional arguments with the same types makes the callsites harder to understand and easier to mess up.

**Review by stuartmorgan — APPROVED** [review]
LGTM with some small nits, but definitely wait for @gaaclarke to review as well.

File: packages/pigeon/lib/dart_generator.dart
Comment (tarrinneal) [reply packages/pigeon/lib/dart_generator.dart:86]: Just seemed like unneeded  code. I thought about it. If you think that's best, I will add it though.

File: packages/pigeon/lib/kotlin_generator.dart
Comment (tarrinneal) [reply packages/pigeon/lib/kotlin_generator.dart:77]: I could do that. The parameter is never used either way though. There's no added safety of adding an assertion atm, and I don't see why that would change at all.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (tarrinneal) [reply packages/pigeon/lib/pigeon_lib.dart:369]: I didn't write any of these, but I can update them.

File: packages/pigeon/lib/dart_generator.dart
Comment (tarrinneal) [reply packages/pigeon/lib/dart_generator.dart:86]: I was considering that, I just didn't think about it when I wrote the code. I'll change it.

File: packages/pigeon/lib/kotlin_generator.dart
Comment (tarrinneal) [reply packages/pigeon/lib/kotlin_generator.dart:77]: I just figured it would be better to have it not fail if it wouldn't have any negative effects regardless. If you disagree that's fine though, I can change it.

Comment (stuartmorgan) [reply packages/pigeon/lib/kotlin_generator.dart:77]: > if it wouldn't have any negative effects regardless

If a generator adapter were accidentally configured with the wrong set of file types it would do duplicate generation, which isn't harmful exactly, but would do a bunch of wasted work. Asserts would make type misconfiguration an obvious error instead of a silent bug.

But it's not a big deal either way.

Comment (tarrinneal) [reply packages/pigeon/lib/kotlin_generator.dart:77]: The actual list of file types doesn't force the generator to use them all. It's still called one at a time.

File: packages/pigeon/lib/dart_generator.dart
Comment (tarrinneal) [reply packages/pigeon/lib/dart_generator.dart:86]: The issue with this, upon thinking about it more, is that it would generate the test files when generating non test dart code as well. I don't know if that is super important to avoid, but it seems unnecessary.

Comment (stuartmorgan) [reply packages/pigeon/lib/dart_generator.dart:86]: We might be able to avoid that by making the set of types option-driven, but let's not worry about it now and just leave it as you have it; we can always revisit later and it's a minor thing.

File: packages/pigeon/lib/kotlin_generator.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/kotlin_generator.dart:77]: > The actual list of file types doesn't force the generator to use them all. It's still called one at a time.

Isn't it driven by a generic for loop that just assumes that it should be called once with each item in the `fileTypeList` though?

File: packages/pigeon/lib/generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/generator.dart:14]: Thanks for making the adjustment.  I think having a Generator interface and maintaining the Adapter keeps the generator files simple.  Everything looks good but one thing.  I think I missed some discussion on `FileType`.  I don't think you've really gained anything by its inclusion.  I think it's a design mistake for the following reasons:

1) Implementing an abstract function and asserting that one parameter is exactly one value means the abstraction doesn't match the implementations well.
1) The `FileType` already has 2 omissions: The AST Generator and the Dart Test generator whose output is not a "source" or a "header"
1) Considering input to some generators being "header" doesn't make sense in many of the generators.
1) The design isn't modular in that new types of output has to be added to a finite list of outputs and all Generators would have to consider what it means to them.

What do we think we gained by adding them?  Now there is a grouping of similar generators: CppHeader and CppSource.  That seems like it complicates the design with little benefit.  What do you think?

Comment (tarrinneal) [reply packages/pigeon/lib/generator.dart:14]: Thanks for the feedback.

Would the inclusion of a `default` option solve some of these issues?

This allows for generators that don't have multiple file types to ignore the `fileType` parameter without mislabeling them. It would also avoid the need to add more options to the FileType enum in the future (unless there is some other multi file system that isn't covered by `source` and `header`).

Is there a better way to handle combining the generator classes without the use of a parameter similar to this implementation?

Comment (gaaclarke) [reply packages/pigeon/lib/generator.dart:14]: The Generator abstraction is already powerful enough to capture the idea of a group.

VP9DJiCm48NtSufHLxJW2qHHQ8LGMI21b5p0YeTYDTYMnqP2qBlZ_BGDeUacSjvxy_IDvJK7wKFyQ02q2UJVP4su9KDU1klpgi1lGBp5NI_HZNL1MyCPLdSeEMIuATE9jTbdMARl4WuNgpPAkYGETVnkOwEsnz9bFtlVUP-oqQfLZ_N9VTOSyaVlhIzcS5xrZgnvwN-leFq3MWbnfWIe6ycC3ywTj8HyE8zX_fC6nZxy2Qzw

or

VP1D3e8m44RtSug9Aq6v02482yF6H1EuG4ChDjQMjCMDUdSB5I7nPzDCt_Hx-TBCMA9jTn40N5gcZwHcM339DB5A9rMADq1SOUCHMwhMSYLDZDKQYR4nvgMR39Vd64jt1l3ugiefQHrywSn9TO8MepJmsSsmknB1QKz7lTlkkB79Lckbqnzr3hnXIkxzzK-rZq9X54qj0Mf1Z9b0eLNNkjBpf6TzRX4kPjdtXCbzBXtU0sHC

I still don't think that the group is buying you much in terms of making the codebase easier to understand or work with though.

Comment (stuartmorgan) [reply packages/pigeon/lib/generator.dart:14]: > What do we think we gained by adding them? Now there is a grouping of similar generators: CppHeader and CppSource.

The issue is that they aren't "similar" they are two fundamentally linked things that can't meaningfully operate on their own. To be useful, each one *must* be paired with a call to the other with identical inputs. And changes to the header generator almost always require exactly corresponding changes to the source generator or everything breaks.

That's why I would like to see combined classes for all of the linked groups; the goal here is to move all of the generator logic into the subclasses, and inherently codependent logic should be in the same class.

(Having two parallel classes, each with half the logic, feels like when someone has two Ivar arrays where the object indexes must match, instead of using a map or an array of structures. Technically it works, but it's really fragile and harder to understand.)

Comment (tarrinneal) [reply packages/pigeon/lib/generator.dart:14]: Working on a design doc for this to allow for more effective communication about our goals here. https://docs.google.com/document/d/1ABr5yZM_sOZTd66bIVpsCJ7_koKmH0tZuLH4ZT1cs8c/edit?usp=sharing

**Comment by tarrinneal** [comment]
closed for https://github.com/flutter/packages/pull/3022

**Pull request closed without merging**
